test/callback/test_diagnostic_hdf5_output.py

```diff
@@ -0,0 +1,154 @@
+"""
+Tests diagnostic callbacks and hdf file output.
+"""
+from thetis import *
+from thetis.callback import VolumeConservation3DCallback
+import h5py
+import shutil
+
+
+def test_callbacks():
+
+    lx = 45000.0
+    ly = 3000.0
+    nx = 25
+    ny = 2
+    mesh2d = RectangleMesh(nx, ny, lx, ly)
+    depth = 50.0
+    elev_amp = 1.0
+    n_layers = 6
+    # estimate of max advective velocity used to estimate time step
+    u_mag = Constant(0.5)
+
+    outputdir = 'outputs'
+    print_info('Exporting to '+outputdir)
+
+    # bathymetry
+    p1_2d = FunctionSpace(mesh2d, 'CG', 1)
+    bathymetry_2d = Function(p1_2d, name='Bathymetry')
+    bathymetry_2d.assign(depth)
+
+    # set time step, export interval and run duration
+    c_wave = float(np.sqrt(9.81*depth))
+    t_cycle = lx/c_wave
+    n_steps = 20
+    dt = round(float(t_cycle/n_steps))
+    t_export = dt
+    t_end = t_cycle/2
+
+    # create solver
+    solver_obj = solver.FlowSolver(mesh2d, bathymetry_2d, n_layers)
+    options = solver_obj.options
+    options.nonlin = False
+    options.solve_salt = True
+    options.solve_temp = False
+    options.solve_vert_diffusion = False
+    options.use_bottom_friction = False
+    options.use_ale_moving_mesh = False
+    if options.use_mode_split:
+        options.dt = dt/5.0
+    else:
+        options.dt = dt/40.0
+    options.t_export = t_export
+    options.t_end = t_end
+    options.u_advection = u_mag
+    options.check_salt_conservation = True
+    options.check_salt_overshoot = True
+    options.check_vol_conservation_2d = True
+    options.check_vol_conservation_3d = True
+    options.fields_to_export = []
+    options.fields_to_export_hdf5 = []
+
+    # need to call creator to create the function spaces
+    solver_obj.create_equations()
+    elev_init = Function(solver_obj.function_spaces.H_2d)
+    elev_init.project(Expression('-eta_amp*cos(2*pi*x[0]/lx)', eta_amp=elev_amp,
+                                 lx=lx))
+    salt_init3d = Constant(3.45)
+
+    class ConstCallback(DiagnosticCallback):
+        """Simple callback example"""
+        name = 'constintegral'
+        variable_names = ['constant', 'integral']
+
+        def __init__(self, const_val, outputdir=None, export_to_hdf5=False,
+                     append_to_log=True):
+            super(ConstCallback, self).__init__(outputdir,
+                                                export_to_hdf5,
+                                                append_to_log)
+            self.const_val = const_val
+
+        def __call__(self, solver_obj):
+            value = self.const_val
+            c = Constant(self.const_val,
+                         domain=solver_obj.mesh.coordinates.ufl_domain())
```

Comment:
I need to explicitly set the domain to avoid error `UFLException: This integral is missing an integration domain.`